gr-utils/python/modtool/core/add.py

```diff
@@ -0,0 +1,299 @@
+                                                                               self.info['blockname'])
+                                    )
+            self.scm.mark_files_updated((self._file['qalib'],))
+        except IOError:
+            logger.warning("Can't add C++ QA files.")
+
+    def _run_lib(self):
+        """ Do everything that needs doing in the subdir 'lib' and 'include'.
+        - add .cc and .h files
+        - include them into CMakeLists.txt
+        - check if C++ QA code is req'd
+        - if yes, create qa_*.{cc,h} and add them to CMakeLists.txt
+        """
+        fname_cc = None
+        fname_h  = None
+        if self.info['version']  == '37':
+            fname_h  = self.info['blockname'] + '.h'
+            fname_cc = self.info['blockname'] + '.cc'
+            if self.info['blocktype'] in ('source', 'sink', 'sync', 'decimator',
+                                           'interpolator', 'general', 'hier', 'tagged_stream'):
+                fname_cc = self.info['blockname'] + '_impl.cc'
+                self._write_tpl('block_impl_h',   'lib', self.info['blockname'] + '_impl.h')
+            self._write_tpl('block_impl_cpp', 'lib', fname_cc)
+            self._write_tpl('block_def_h',    self.info['includedir'], fname_h)
+        else: # Pre-3.7 or autotools
+            fname_h  = self.info['fullblockname'] + '.h'
+            fname_cc = self.info['fullblockname'] + '.cc'
+            self._write_tpl('block_h36',   self.info['includedir'], fname_h)
+            self._write_tpl('block_cpp36', 'lib',                    fname_cc)
+        if self.add_cc_qa:
+            if self.info['version'] == '37':
+                self._run_cc_qa()
+            elif self.info['version'] == '36':
+                logger.warning("Warning: C++ QA files not supported for 3.6-style OOTs.")
+            elif self.info['version'] == 'autofoo':
+                logger.warning("Warning: C++ QA files not supported for autotools.")
+        if not self.skip_cmakefiles:
+            ed = CMakeFileEditor(self._file['cmlib'])
+            cmake_list_var = '[a-z]*_?' + self.info['modname'] + '_sources'
+            if not ed.append_value('list', fname_cc, to_ignore_start='APPEND ' + cmake_list_var):
+                ed.append_value('add_library', fname_cc)
+            ed.write()
+            ed = CMakeFileEditor(self._file['cminclude'])
+            ed.append_value('install', fname_h, to_ignore_end='DESTINATION[^()]+')
+            ed.write()
+            self.scm.mark_files_updated((self._file['cminclude'], self._file['cmlib']))
+
+    def _run_swig(self):
+        """ Do everything that needs doing in the subdir 'swig'.
+        - Edit main *.i file
+        """
+        if self._get_mainswigfile() is None:
+            logger.warning('Warning: No main swig file found.')
+            return
+        logger.info("Editing %s..." % self._file['swig'])
```

Comment:
Using `'{}'.format(foo)` is preferred for string formatting

Reply:
Oh! Yes, I just changed from `print` to `logging.logger`. I will amend it though.